scipy/stats/qmc.py

```diff
@@ -0,0 +1,648 @@
+
+    return primes
+
+
+def van_der_corput(n_samples, base=2, start_index=0):
+    """Van der Corput sequence.
+
+    Pseudo-random number generator based on a b-adic expansion.
+
+    Parameters
+    ----------
+    n_samples : int
+        Number of element of the sequence.
+    base : int
+        Base of the sequence.
+    start_index : int
+        Index to start the sequence from.
+
+    Returns
+    -------
+    sequence : list (n_samples,)
+        Sequence of Van der Corput.
+
+    """
+    sequence = []
+    for i in range(start_index, start_index + n_samples):
+        n_th_number, denom = 0., 1.
+        quotient = i
+        while quotient > 0:
+            quotient, remainder = divmod(quotient, base)
+            denom *= base
+            n_th_number += remainder / denom
+        sequence.append(n_th_number)
+
+    return sequence
+
+
+def halton(dim, n_samples, bounds=None, start_index=0):
+    """Halton sequence.
+
+    Pseudo-random number generator that generalize the Van der Corput sequence
+    for multiple dimensions. Halton sequence use base-two Van der Corput
+    sequence for the first dimension, base-three for its second and base-n for
+    its n-dimension.
+
+    Parameters
+    ----------
+    dim : int
+        Dimension of the parameter space.
+    n_samples : int
+        Number of samples to generate in the parameter space.
+    bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+        Desired range of transformed data. The transformation apply the bounds
+        on the sample and not the theoretical space, unit cube. Thus min and
+        max values of the sample will coincide with the bounds.
+    start_index : int
+        Index to start the sequence from.
+
+    Returns
+    -------
+    sequence : array_like (n_samples, k_vars)
+        Sequence of Halton.
+
+    References
+    ----------
+    [1] Halton, "On the efficiency of certain quasi-random sequences of points
+      in evaluating multi-dimensional integrals", Numerische Mathematik, 1960.
+
+    Examples
+    --------
+    Generate samples from a low discrepancy sequence of Halton.
+
+    >>> from scipy.stats import qmc
+    >>> sample = qmc.halton(dim=2, n_samples=5)
+
+    Compute the quality of the sample using the discrepancy criterion.
+
+    >>> uniformity = qmc.discrepancy(sample)
+
+    If some wants to continue an existing design, extra points can be obtained.
+
+    >>> sample_continued = qmc.halton(dim=2, n_samples=5, start_index=5)
+
+    """
+    base = n_primes(dim)
+
+    # Generate a sample using a Van der Corput sequence per dimension.
+    sample = [van_der_corput(n_samples + 1, bdim, start_index) for bdim in base]
+    sample = np.array(sample).T[1:]
+
+    # Sample scaling from unit hypercube to feature range
+    if bounds is not None:
+        min_ = bounds.min(axis=0)
+        max_ = bounds.max(axis=0)
+        sample = sample * (max_ - min_) + min_
+
+    return sample
+
+
+global best_doe, best_disc
```

Comment:
please avoid using globals

Reply:
Hmmm at the time of writing this code. We did not really see how to better do this. If you have any advice, I will be happy to avoid this as well 👍

Reply:
Resolve with the class-based API.